mrmustard/math/tensor_networks/tensors.py

```diff
@@ -0,0 +1,328 @@
+from dataclasses import dataclass, field
+from typing import List, Optional
+
+import numpy as np
+import uuid
+
+
+__all__ = ["Wire", "Tensor"]
+
+
+def random_int() -> int:
+    r"""
+    Returns a random integer obtained from a UUID
+    """
+    return uuid.uuid1().int
+
+
+@dataclass
+class Wire:
+    r"""Represents a wire in a tensor network.
+
+    Each wire is characterized by a unique identifier ``id``, which must different from
+    the identifiers of all the other wires in the tensor network, as well as by a label
+    ``mode`` that represents a mode of light.
+
+    Args:
+        id: A numerical identifier for this wire.
+        mode: The mode represented by this wire.
+        is_input: Whether this wire is an input to a tensor or an output.
+        is_ket: Whether this wire is on the ket or on the bra side.
+
+    """
+    id: int
+    mode: int
+    is_input: bool
+    is_ket: bool
+
+    def __post_init__(self):
+        self._contraction_id: int = random_int()
+        self._is_connected = False
+
+    @property
+    def contraction_id(self) -> int:
+        r"""
+        A numerical identifier for the contraction involving this wire.
+        """
+        return self._contraction_id
+
+    @contraction_id.setter
+    def contraction_id(self, value: int):
+        self._contraction_id = value
+
+    @property
+    def is_connected(self) -> bool:
+        r"""
+        Whether or not this wire is connected with another wire.
+        """
+        return self._is_connected
+
+    @is_connected.setter
+    def is_connected(self, value: bool):
+        self._is_connected = value
+
+
+@dataclass
+class WireGroup:
+    r"""A group of wires in a tensor network.
+
+    Args:
+        ket: A dictionary containing the wires on the ket side.
+        bra: A dictionary containing the wires on the bra side.
+
+    """
+    ket: dict = field(default_factory=dict)
+    bra: dict = field(default_factory=dict)
+
+
+class Tensor(ABC):
+    r"""An abstract class representing a tensor in a tensor network.
+
+    In Mr Mustard, tensors are used to represent a state or a transformation on a given set
+    of modes in the Fock representation. For example, a single-mode unitary matrix
+    :math:`U=\sum_{i,j=1}^Nu_{i,j}|i\rangle\langle{j}|` acting on mode ``3`` in an
+    N-dimensional Fock basis corresponds to the following ``Tensor`` object:
+
+    .. code-block::
+        class U(Tensor):
+            def value(self, cutoff):
+                # specify the value of the tensor
+                pass
+
+        U("my_unitary", [3], [3], [3], [3])
+
+    Args:
+        name: The name of this tensor.
+        modes_in_ket: The input modes on the ket side.
+        modes_out_ket: The output modes on the ket side.
+        modes_in_bra: The input modes on the bra side.
+        modes_out_bra: The output modes on the bra side.
+    """
+    _repr_markdown_ = None  # otherwise it takes over the repr due to mro
+
+    def __init__(
```

Comment:
should `ABC`s have an init? 🤔

Reply:
I never really added it, so I'd lean towards saying no. Can't find a final answer on internet to this though, although there's no super().init used in any examples provided by the official python docs:
https://docs.python.org/3/library/abc.html